dist/clouduploader.rb

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/ruby
+require 'json'
+require 'ostruct'
+require 'open3'
+
+if ARGV.length != 5
+  raise 'Wrong number of arguments, please provide: user platform upload_file targetdata filename'
+end
+
+platform = ARGV[1]
+image_path = ARGV[2]
+data_path = ARGV[3]
+filename = ARGV[4]
+
+data = JSON.parse(File.read(data_path))
+DURATION = 60 * 60
+KEY_NAME = 'obs'.freeze
+PRIVATE_KEY = '/etc/clouduploader.pem'.freeze
+HOME = '/etc/obs/cloudupload'.freeze
+ENV['HOME'] = HOME
+ENV['PYTHONUNBUFFERED'] = '1'
+
+def get_ec2_credentials(arn, external_id)
+  # Credentials are stored in  ~/.aws/credentials
+  out, err, status = Open3.capture3(
+    'aws',
+    'sts',
+    'assume-role',
+    "--role-arn=#{arn}",
+    "--external-id=#{external_id}",
+    '--role-session-name=obs',
+    "--duration-seconds=#{DURATION}"
+  )
+
+  if status.success?
+    STDOUT.write("Successfully authenticated.\n")
+    json = JSON.parse(out)
+    OpenStruct.new(
+      access_key_id: json['Credentials']['AccessKeyId'],
+      secret_access_key: json['Credentials']['SecretAccessKey'],
+      session_token: json['Credentials']['SessionToken']
+    )
+  else
+    abort(err)
+  end
+end
+
+def upload_image_to_ec2(image, credentials, region, filename)
+  STDOUT.write("Start uploading image #{filename}.\n")
+  name = File.basename(filename)
+
+  Open3.popen2e(
+    'ec2uploadimg',
+    "--description='obs uploader'",
+    '--machine=x86_64',
```

Comment:
do we have a failure handling somewhere when it is not a x86_64 image?

Reply:
We do not know the architecture at this point! I would check the architecture in the frontend and only allow x86_64 instead.